docs/advanced/fields.rst

```diff
@@ -0,0 +1,367 @@
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you want to a Page Object with async fields without calling its
+``to_item`` method, make sure to await the field when needed, and
+not await it when that's not needed:
+
+.. code-block:: python
+
+    page = MyPage(...)
+    name = page.name
+    price = await page.price
+
+This is not ideal, because now the code which needs to use a page object
+must be aware if a field is sync or async. If a field needs to be changed
+from being sync to ``async def`` (or the other way around),
+e.g. because of a website change, all the code which uses this page
+object must be updated.
+
+One approach to solve it is to always define all fields as ``async def``.
+It works, but it makes the page objects harder to use in non-async environments.
+
+Instead of doing this, you can also use :func:`~.ensure_awaitable` utility
+function when accessing the fields:
+
+.. code-block:: python
+
+    from web_poet.utils import ensure_awaitable
+
+    page = MyPage(...)
+    name = await ensure_awaitable(page.name)
+    price = await ensure_awaitable(page.price)
+
+Now any field can be converted from sync to async, or the other way around,
+and the code would keep working.
+
+Item classes
+------------
+
+In all previous examples, ``to_item`` methods are returning ``dict``
+instances. It is common to use item classes (e.g. dataclasses or
+attrs instances) instead of unstructured dicts to hold the data:
+
+.. code-block:: python
+
+    import attrs
+    from web_poet import ItemPage, HttpResponse
+
+    @attrs.define
+    class Item:
+        name: str
+        price: str
+
+
+    @attrs.define
+    class MyPage(ItemPage):
+        # ...
+        def to_item(self) -> Item:
+            return Item(
+                name=self.name,
+                price=self.price
+            )
+
+:mod:`web_poet.fields` supports it, by allowing to pass an item class to the
+:func:`~.item_from_fields` / :func:`~.item_from_fields_sync` functions:
+
+.. code-block:: python
+
+    @attrs.define
+    class MyPage(ItemPage):
+        # ...
+
+        async def to_item(self) -> Item:
+            return await item_from_fields(self, item_cls=Item)
+
+
+This approach plays particularly well with the
+:func:`@field <web_poet.fields.field>` decorator, preventing some of the errors,
+which may happen if results are plain "dicts".
+
+Consider the following badly written page object:
+
+.. code-block:: python
+
+    import attrs
+    from web_poet import ItemPage, HttpResponse, field, item_from_fields
+
+    @attrs.define
+    class Item:
+        name: str
+        price: str
+
+
+    @attrs.define
+    class MyPage(ItemPage):
+        response: HttpResponse
+
+        @field
+        def nane(self):
+            return self.response.css(".name").get()
+
+        async def to_item(self) -> Item:
+            return await item_from_fields(self, item_cls=Item)
+
+Because Item class is used, a typo ("nane" instead of "name") is detected
```

Comment:
Love the idea of detecting missing required arguments, but I'm **not sure about raising errors for fields not present in the Items**. However, I might be biased by my approach of seeing POs as field "providers"...

Take this Item+PO:

```python
@attrs.define
class Item:
    name: str
    price: str


@attrs.define
class DaaSProduct(ItemPage):

    @field
    def name(self):
        return self.response.css(".name").get()

    @field
    def price(self):
        return self.response.css(".price").get()

    async def to_item(self) -> Item:
        return await item_from_fields(self, item_cls=Item)

```

Let's say we want to create a custom PO subclassing it in order to reuse some fields:

```python
@attrs.define
class ClientItem:
    name: str
    description: str


@attrs.define
class ClientProduct(DaaSProduct):

    @field
    def description(self):
        return self.response.css(".price").get()

    async def to_item(self) -> ClientItem:
        return await item_from_fields(self, item_cls=ClientItem)
```
An error will be raised for `price` field here, as is not present in `ClientItem` definition.

Reply:
That's a great point @gatufo, I haven't considered it.

I'm unsure about the default, but it looks like we need to provide an option to skip the fields not present in item. In order to do so, we need to get a list of item fields; I can't find a way around it, if we want to handle required fields as well.

Itemadapter has a method to get item fields, but it works on instances, not on classes. So, it seems the way to go is to implement `ItemAdapter.get_field_names_from_class`; then, modify `item_from_fields_sync` function like this:

```py
def item_from_fields_sync(obj, item_cls=dict, only_item_fields=False):
    """Synchronous version of :func:`item_from_fields`."""
    page_field_names = list(getattr(obj, "_auto_item_fields", {}))
    if only_item_fields:
        try:
            item_field_names = itemadapter.ItemAdapter.get_field_names_from_class(item_cls)
            # fixme: preserve field order
            field_names = set(page_field_names) | set(item_field_names)
        except itemadapter.FieldNamesUnknown:  # e.g. when item_cls is dict
            field_names = page_field_names
    else:
        field_names = page_field_names
    return item_cls(**{name: getattr(obj, name) for name in field_names})

```

Reply:
PR to itemadapter: https://github.com/scrapy/itemadapter/pull/64

Reply:
It is partially addressed here: https://github.com/scrapinghub/web-poet/pull/53/commits/80473c2a732db44a7d24ab9945dae3eb003da0a2. TODO: tests.

Reply:
@gatufo it's ready for another round of review! The most important piece to check is the documentation - "Item classes" section and its sub-sections, especially "Changing Item type".

Don't use the commit linked above, there are some changes since then.